### ISSUE TITLE
We're frequently exceeding our RAM quota on production.

### ISSUE THREAD

**Issue by pozorvlak** [Growstuff/growstuff#1640]
![growstuff_ram](https://user-images.githubusercontent.com/225991/39879830-c8c221c4-5473-11e8-9d45-b0a1e84d6bd1.PNG)

Here's [Heroku's guide to optimizing memory usage for Ruby](https://devcenter.heroku.com/articles/ruby-memory-use).

**Comment by Br3nda** [comment]
Do we know which page?

**Comment by Br3nda** [comment]
Might be  #1638

**Comment by pozorvlak** [comment]
They could definitely be related, but I think the causality might go the other way - exceeding our RAM quota means we're hitting swap, which will slow page-loads way down. Optimising queries may very well help, but adding caching is unlikely to reduce our RAM usage :-)

**Comment by Br3nda** [comment]
i didn't think heroku gave you swap - in my experience it kills the app when you exceed the allocated resources you've paid for.

**Comment by pmackay** [comment]
I notice the whole app runs on one dyno - would using a worker be possible? Or would that create a budget problem?

**Comment by Br3nda** [comment]
i think the individual dyno would still exceed it's RAM quota.

There is a new relic - there may be clues on there on which pages/code is being memory hungry

**Comment by apdarr** [comment]
@Br3nda That's right, on the [Common Runtime](https://devcenter.heroku.com/articles/dynos#common-runtime-networking), where the app runs, dynos can use swap when exceeding memory usage. Hobby dynos run on multitenant machines, which can sometimes mean performance greater than expected if the host is underutilized by competing Heroku apps. 

 @pmackay If the app has or begins to emit [H12 timeout errors](https://devcenter.heroku.com/articles/error-codes#h12-request-timeout) on Heroku, then it's likely time to think about adding a background process if budget allows. Using something like [rack-timeout](https://github.com/heroku/rack-timeout) can also provide an interim reprieve. [The Scout](https://elements.heroku.com/addons/scout) add-on is free and is specifically designed for identifying memory issues and N+1 queries in Rails apps. That combined with New Relic should provide a decent start. 

There are a few other approaches to tackle memory issues, but I thought I'd start with the above info!

**Comment by pozorvlak** [comment]
@apdarr Thanks!

**Comment by Br3nda** [comment]
The biggest memory user so far is the comfy Mexican sofa (on staging). Let's check again in a week

**Comment by Br3nda** [comment]
Scout reports our biggest memory use is consistently in Comfy::Cms::ContentController#show

**Comment by Br3nda** [comment]
Perhaps more interesting is the 2s to 5s response times for HarvestsController#index

**Comment by Br3nda** [comment]
this is still happening. the crops hierarchy page is the current biggest memory user

**Comment by pozorvlak** [comment]
The crops hierarchy page ought to be mostly cached - I wonder if the cache is expiring, or if we're somehow not hitting it? But yeah, I can see how it would be heavy - we call `Crop.toplevel` in the controller, then `crop.varieties` recursively in the view, so it's going to be doing a *lot* of querying. Is there some way we can pre-populate the `varieties` field in the initial query? Or fetch `Crop.all` in the controller and construct the varieties tree in memory? Or maybe we could show all the top-level crops collapsed by default and fetch them lazily using AJAX?

**Comment by Br3nda** [comment]
This gem should give use some useful caching for descendants
https://github.com/stefankroes/ancestry

**Comment by Br3nda** [comment]
Fixed memcache at 9am. It's been looking good since. 
![Screenshot_20191207_154015](https://user-images.githubusercontent.com/12266/70367774-4c5fc900-1908-11ea-98e3-cdc1d638a316.jpg)
![Screenshot_20191207_153939](https://user-images.githubusercontent.com/12266/70367775-4cf85f80-1908-11ea-9e30-e6807d9bf294.jpg)
![Screenshot_20191207_153954](https://user-images.githubusercontent.com/12266/70367776-4d90f600-1908-11ea-86c7-61ef06d413e8.jpg)

**Comment by pozorvlak** [comment]
Awesome!

**Comment by Br3nda** [comment]
I'd like to get it under 500M before closing this

**Comment by Br3nda** [comment]
![Screenshot_20191209_145024](https://user-images.githubusercontent.com/12266/70401152-4ea25e80-1a93-11ea-8d22-01d318368a99.jpg)

**Comment by Br3nda** [comment]
there's a memory leak somewhere

Meanwhile the home page is now very very fast. I'll go through the other major pages and gets them nice as well.

**Comment by Br3nda** [comment]
Still sitting at 700MB

![Screenshot_20191214_100518](https://user-images.githubusercontent.com/12266/70832188-5417f400-1e59-11ea-9ed2-07a0c46fb964.jpg)

**Comment by Br3nda** [comment]
i think we've got it under 500Mbyte with the deployment of #2320 . Will watch it for a few more days
![image](https://user-images.githubusercontent.com/12266/72304426-604fd500-36d5-11ea-90b6-e62b6b92b0d3.png)

**Comment by CloCkWeRX** [comment]
Has crept back up a bit:
![image](https://github.com/Growstuff/growstuff/assets/365751/3af813ab-52f0-4d9f-8b18-d9137ea1c076)

**Comment by CloCkWeRX** [comment]
Recent rounds of work:
- Fix a bunch of routes where if a member or planting was missing, it would simply show all records like /photos; causing bots to crawl endlessly
- Banned Semrush, which was relentlessly crawling is, via robots.txt
- Deleted many empty members with only the default garden and no other activity in the last three years.
- More caching on various areas
- Fixed busted links with an ?owner=xyz that didn't actually filter, to discourage crawling

Far from perfect, but cleaned up some of the memory use growth from newer features; ie:


<img width="1740" height="710" alt="Image" src="https://github.com/user-attachments/assets/dafec85b-2dad-463d-b1db-e16ff88eec0e" />

**Comment by CloCkWeRX** [comment]
Misc AI suggested areas to optimise:


Memory optimisation review — Growstuff

- [ ] Garden chart/timeline loads all historical plantings — app/controllers/charts/gardens_controller.rb:7-15. Bound by date window and pluck instead of instantiating models.
- [ ] Garden show/index render all plantings and photos — app/views/gardens/show.html.haml:52,184-188, app/views/gardens/_card.html.haml. Paginate photos, .limit progress lists, or render counts.
- [ ] initialize_on_precompile = true contradicts its comment — config/application.rb:56. Forces full app boot (and DB connect) during asset precompile.
- [ ] Crop#update_medians iterates every planting — app/models/crop.rb:148; Garden#mark_inactive... saves each planting — app/models/garden.rb:78-81. Use in_batches/update_all.
- [ ] harvest_reminder Ruby-filters all active plantings — app/mailers/notifier_mailer.rb:62, lib/tasks/growstuff.rake:64-66. select(&:predicate) loads the entire relation; push the predicate into SQL and use exists?.
- [ ] Unbounded per-request memoization — app/helpers/crops_helper.rb:5,26, app/helpers/plantings_helper.rb:49, app/models/crop.rb:167-173 grow with every record rendered on large index/export pages.
- [ ] places#index serializes every located member — app/controllers/places_controller.rb:12-14 and :27-29. Member.located.to_json materialises the whole table plus one giant JSON string. Paginate, pluck only the six fields, and .limit the nearest_to result.
- [x] https://github.com/Growstuff/growstuff/pull/4829
- [x] https://github.com/Growstuff/growstuff/pull/4809
- [x] https://github.com/Growstuff/growstuff/pull/4810
- [x] https://github.com/Growstuff/growstuff/pull/4811
- [x] https://github.com/Growstuff/growstuff/pull/4812
- [x] https://github.com/Growstuff/growstuff/pull/4806
- [x] https://github.com/Growstuff/growstuff/pull/4807
- [x] https://github.com/Growstuff/growstuff/pull/4805
- [x] https://github.com/Growstuff/growstuff/pull/4796
- [x] https://github.com/Growstuff/growstuff/pull/4797
- [x] https://github.com/Growstuff/growstuff/pull/4799
- [x] https://github.com/Growstuff/growstuff/pull/4791
- [x] https://github.com/Growstuff/growstuff/pull/4790
- [x] https://github.com/Growstuff/growstuff/pull/4793
